Signed-off-by: Jussi Kukkonen <jussi.kukkonen@intel.com>
Upstream-Status: Backport [commit 59a731be6b in '1.10' branch]


From 02a6002c3a80b3a5301c0943b1a1518bbdf439fc Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?V=C3=ADctor=20Manuel=20J=C3=A1quez=20Leal?=
 <vjaquez@igalia.com>
Date: Fri, 21 Apr 2017 19:07:18 +0200
Subject: [PATCH] vaapivideobufferpool: create allocator if needed

Backport to branch 1.10

Sometimes a video decoder could set different buffer pool
configurations, because their frame size changes. In this case we
did not reconfigure the allocator.

This patch enables this use case, creating a new allocator inside
the VAAPI buffer pool if the caps changed, if it is not dmabuf-based.
If so, it is just reconfigured, since it doesn't have a surface pool.

https://bugzilla.gnome.org/show_bug.cgi?id=781577
---
 gst/vaapi/gstvaapivideobufferpool.c | 21 +++++++++++++++++++++
 1 file changed, 21 insertions(+)

diff --git a/gst/vaapi/gstvaapivideobufferpool.c b/gst/vaapi/gstvaapivideobufferpool.c
index a3b9223f..9a50614b 100644
--- a/gst/vaapi/gstvaapivideobufferpool.c
+++ b/gst/vaapi/gstvaapivideobufferpool.c
@@ -159,6 +159,27 @@ gst_vaapi_video_buffer_pool_set_config (GstBufferPool * pool,
     gst_object_replace ((GstObject **) & priv->allocator, NULL);
   priv->video_info = new_vip;
 
+  {
+    guint surface_alloc_flags;
+    gboolean vinfo_changed = FALSE;
+
+    if (allocator) {
+      const GstVideoInfo *alloc_vinfo =
+          gst_allocator_get_vaapi_video_info (allocator, &surface_alloc_flags);
+      vinfo_changed = gst_video_info_changed (alloc_vinfo, &new_vip);
+    }
+
+    if (vinfo_changed && allocator && priv->use_dmabuf_memory) {
+      gst_allocator_set_vaapi_video_info (allocator, &new_vip,
+          surface_alloc_flags);
+    } else if (!priv->use_dmabuf_memory && (vinfo_changed || !allocator)) {
+      /* let's destroy the other allocator and create a new one */
+      allocator = gst_vaapi_video_allocator_new (priv->display, &new_vip,
+          surface_alloc_flags);
+      gst_buffer_pool_config_set_allocator (config, allocator, NULL);
+    }
+  }
+
   if (!gst_buffer_pool_config_has_option (config,
           GST_BUFFER_POOL_OPTION_VAAPI_VIDEO_META))
     goto error_no_vaapi_video_meta_option;
-- 
2.11.0


